contributors/design-proposals/storage/online-grow-volume-size.md

```diff
@@ -0,0 +1,344 @@
+                uniqueVolumeName, pod.Name, pod.Namespace)
+            continue
+        }
+
+        // If volume hasn't been mounted to the node, we do not add annotation to pods running on this node.
+        if !mountedNodes[podObj.Spec.NodeName] {
+            glog.V(5).Infof("Skip MarkForFileSystemResize for pvc %s pod %s/%s: "+
+                "volume not mounted to %s yet", uniqueVolumeName, pod.Name, pod.Namespace, podObj.Spec.NodeName)
+            continue
+        }
+
+        // Check whether one or more pods has been updated in the
+        // node or not, if so, skip other pods running on same node.
+        if markedNodes[commontypes.NodeName(podObj.Spec.NodeName)] {
+            continue
+        }
+
+        if resizeMap.podHasMarkedForPVCResizing(podObj, volumeSpec.Name()) {
+            markedNodes[commontypes.NodeName(podObj.Spec.NodeName)] = true
+            glog.V(5).Infof("Skip update pod %s/%s annotation for "+
+                "pvc %s resizing as it already updated", podObj.Name, podObj.Namespace, uniqueVolumeName)
+        } else {
+            podObjs = append(podObjs, podObj)
+        }
+    }
+
+    for _, podObj := range podObjs {
+        if markedNodes[commontypes.NodeName(podObj.Spec.NodeName)] {
+            glog.V(5).Infof("Skip MarkForFileSystemResize for pvc %s pod %s/%s: "+
+                "another pod in the same node has been marked", uniqueVolumeName, podObj.Name, podObj.Namespace)
+            continue
+        }
+
+        markErr := resizeMap.updatePodAnnotationForPVCResizing(podObj, volumeSpec.Name())
+        if markErr != nil {
+            glog.Errorf("Update pod %s/%s annotation for pvc %s resizing failed: %v", podObj.Name, podObj.Namespace, uniqueVolumeName, markErr)
+            return markErr
+        }
+        markedNodes[commontypes.NodeName(podObj.Spec.NodeName)] = true
+    }
+
+    return nil
+}
+```
+
+### File System Resize on Kubelet
+
+#### Prerequisite of Online File System Resize
+
+* Pod running on the node has one more `volumeFSResizingAnnotation` annotations.
+* Volume with a online resizing request has been mounted to the node.
+* `PV.Spec.Capacity` is greater than `PVC.Status.Capacity`.
+
+#### Volume Resize Request Detect
+
+Kubelet need to periodically loop through the list of active pods to detect which volume requires file system resizing. We use a `VolumeToResize` object to represent a volume resizing requests and introduce a new interface `VolumeFsResizeMap` rather than `DesiredStateOfWorld` to cache such `VolumeToResize` objects, since we treat it as a request queue, not a state.
+
+Besides, we add a `VolumeFsResizeMapPopulator` interface as the producer of `VolumeFsResizeMap`, it runs an asynchronous periodic loop to populate the volumes need file system resize. In each loop, `VolumeFsResizeMapPopulator` traverses each pod's annotation, and generate a `VolumeToResize` object for each `volumeFSResizingAnnotation`.
```

Comment:
Will this be a new package or part of volume-manager's interface?

Reply:
I'm not sure if I understand what you mean. We can put `VolumeFsResizeMap` interface under package `pkg/kubelet/volumemanager/cache`, just like the ASW and DSW, and put `VolumeFsResizeMapPopulator` interface under package `pkg/kubelet/volumemanager/populator`, just like the `DesiredStateOfWorldPopulator`. Then `pkg/kubelet/volumemanager/reconciler.Reconciler` can work as a consumer of `VolumeFsResizeMap`. These two interfaces will look like this:
```go
type VolumeFsResizeMap interface {
    AddVolumeForPod(volumeName string, pod *v1.Pod)
    GetVolumesToResize() []VolumeToResize
    MarkAsFileSystemResized(pod *v1.Pod, volumeName string) error
}
```

```go
type VolumeFsResizeMapPopulator interface {
    Run(stopCh <-chan struct{})
}
```